frontend/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -14,465 +14,300 @@
-        if (sort.direction === "asc") {
-          // Sort column descending
-          sortDirection = "desc"
-        } else {
-          // Remove sorting of column
-          setSort(undefined)
-          return
-        }
+  const [numRows, setNumRows] = React.useState(
+    editingState.current.getNumRows()
+  )
+
+  React.useEffect(() => {
+    editingState.current = new EditingState(originalNumRows)
+    setNumRows(editingState.current.getNumRows())
+  }, [originalNumRows])
+
+  const resetEditingState = React.useCallback(() => {
+    editingState.current = new EditingState(originalNumRows)
+    setNumRows(editingState.current.getNumRows())
+  }, [originalNumRows])
+
+  const { columns: originalColumns, getCellContent: getOriginalCellContent } =
+    useDataLoader(element, data, numRows, disabled, editingState)
+
+  const { columns, sortColumn, getOriginalIndex, getCellContent } =
+    useColumnSort(originalNumRows, originalColumns, getOriginalCellContent)
+
+  /**
+   * This callback should be called after any edits have been applied to the data.
+   * It will finish up the editing by updating the number of rows, clearing the selection,
+   * and triggering a rerun of the script.
+   *
+   * @param clearSelections - Whether to clear the selection. This is usually done after deleting rows.
+   * @param triggerRerun - Whether to trigger a rerun of the script after applying edits
+   */
+  const applyEdits = React.useCallback(
+    (clearSelections = false, triggerRerun = true) => {
+      if (numRows !== editingState.current.getNumRows()) {
+        // Reset the number of rows if it has been changed in the editing state
+        setNumRows(editingState.current.getNumRows())
+      }
+
+      if (clearSelections) {
+        clearSelection()
       }
 
-      setSort({
-        column: clickedColumn,
-        direction: sortDirection,
-        mode: getColumnSortMode(
-          (clickedColumn as GridColumnWithCellTemplate).columnType
-        ),
-      } as ColumnSortConfig)
+      if (triggerRerun) {
+        // Use debounce to prevent rapid updates to the widget state.
+        debounce(DEBOUNCE_TIME_MS, () => {
+          const currentEditingState = editingState.current.toJson(columns)
+          let currentWidgetState = widgetMgr.getStringValue(
+            element as WidgetInfo
+          )
+
+          if (currentWidgetState === undefined) {
+            // Create an empty widget state
+            currentWidgetState = new EditingState(0).toJson([])
+          }
+
+          // Only update if there is actually a difference between editing and widget state
+          if (currentEditingState !== currentWidgetState) {
+            widgetMgr.setStringValue(
+              element as WidgetInfo,
+              currentEditingState,
+              {
+                fromUi: true,
+              }
+            )
+          }
+        })()
+      }
     },
-    [sort, columns]
+    [widgetMgr, element, numRows]
```

Comment:
Does widgetMgr need to be included in here?

Reply:
I will leave this in for now since it works fine for now and could have hard to test downsides.  But will look at this again as a fast-follow refactoring after the release.